server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java

```diff
@@ -246,50 +258,93 @@ public Invocations getInvocations() {
+            invocationHelper.getExecutor()))
+            .orElseGet(() -> completedFuture(responseBuilder.build()));
 
-      return CompletableFuture.supplyAsync(
-            () -> {
-               Map<Address, String> result = Security.doAs(request.getSubject(), (PrivilegedAction<Map<Address, String>>) () -> xsiteAdmin.nodeStatus(site));
-               return addEntityAsJson(Json.make(result), responseBuilder).build();
-            }
-            , invocationHelper.getExecutor());
    }
 
    private <T> CompletionStage<RestResponse> statusOperation(RestRequest request, Function<XSiteAdminOperations, T> op) {
       NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
-      XSiteAdminOperations xsiteAdmin = getXSiteAdmin(request);
-
-      return CompletableFuture.supplyAsync(
+      Optional<XSiteAdminOperations> xsiteAdmin = getXSiteAdmin(request, responseBuilder);
+      return xsiteAdmin.<CompletionStage<RestResponse>>map(ops -> supplyAsync(
             () -> {
-               T result = Security.doAs(request.getSubject(), op, xsiteAdmin);
+               T result = Security.doAs(request.getSubject(), (PrivilegedAction<T>) () -> op.apply(ops));
                return addEntityAsJson(Json.make(result), responseBuilder).build();
             },
-            invocationHelper.getExecutor());
+            invocationHelper.getExecutor()))
+            .orElseGet(() -> completedFuture(responseBuilder.build()));
+   }
+
+   private CompletionStage<RestResponse> getStateTransferMode(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      //check if site exists
+      final String site = request.variables().get("site");
+      Optional<XSiteAdminOperations> xsiteAdminOpt = getXSiteAdminAndCheckSite(request, responseBuilder);
+
+      return xsiteAdminOpt.<CompletionStage<RestResponse>>map(ops ->
+             ops.asyncGetStateTransferMode(site).thenApply(s -> addEntityAsJson(Json.make(s), responseBuilder).build()))
+            .orElseGet(() -> completedFuture(responseBuilder.build()));
    }
 
-   private XSiteAdminOperations getXSiteAdmin(RestRequest request) {
+   private CompletionStage<RestResponse> setStateTransferMode(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+      //parse content
+      String mode = request.getParameter("mode");
+      if (mode == null) {
+         return completedFuture(responseBuilder.status(BAD_REQUEST).build());
+      }
+      //check if site exists
+      final String site = request.variables().get("site");
+      Optional<XSiteAdminOperations> xsiteAdminOpt = getXSiteAdminAndCheckSite(request, responseBuilder);
+      return xsiteAdminOpt.<CompletionStage<RestResponse>>map(ops ->
+            ops.asyncSetStateTransferMode(site, mode)
+                  .thenApply(ok -> responseBuilder.status(ok ? OK : NOT_MODIFIED).build()))
```

Comment:
Is `NOT_MODIFIED` sufficient? Or should we include a message explaining that the requested mode is not supported on the remote site?

Reply:
Our REST client ignores the body if NOT_MODIFIED is returned and I'm hoping the users read the docs before invoking commands randomly.